v1_21_4/src/main/java/me/aleksilassila/litematica/printer/v1_21_4/guides/placement/GeneralPlacementGuide.java

```diff
@@ -302,37 +302,62 @@ public PrinterPlacementContext getContextByRotation(ClientPlayerEntity player) {
-        PrinterPlacementContext noRotateContext = new PrinterPlacementContext(player, noRotateHitResult, requiredItem, slot, null, false);
-        BlockState noRotateResult = getRequiredItemAsBlock(player)
-                .orElse(targetState.getBlock())
-                .getPlacementState(noRotateContext); // FIXME torch shift clicks another torch and getPlacementState is the clicked block, which is true
-
-        if (noRotateResult != null && correctObserverPlacement(targetState, noRotateResult) && (statesEqual(noRotateResult, targetState))) {
-            contextCache = noRotateContext;
-            noRotateContext.isAirPlace = true;
-            return noRotateContext;
+        // We'll brute force: sides (including horizontal) + hit positions inside the target block volume
+        // to emulate realistic clicks that yield correct orientation for special blocks (hoppers, froglights, basalt, slabs).
+
+        // Candidate sides to emulate a supporting face. For airplace we pretend there is a block on that side.
+        Direction[] candidateSides = new Direction[]{
+                Direction.UP, Direction.DOWN,
+                Direction.NORTH, Direction.SOUTH, Direction.EAST, Direction.WEST
+        };
+
+        // Hit offsets inside block relative to its min corner (0..1). Include lower y (<0.5) for bottom slabs, and center variations.
+        double[] ySamples = new double[]{0.25, 0.5, 0.75};
+        double[] xzSamples = new double[]{0.25, 0.5, 0.75};
+
+        // First quick attempt: original center (UP side) (fast path)
+        {
+            BlockHitResult hr = new BlockHitResult(Vec3d.ofCenter(state.blockPos), Direction.UP, state.blockPos, true);
+            PrinterPlacementContext quick = new PrinterPlacementContext(player, hr, requiredItem, slot, null, false);
+            BlockState res = getRequiredItemAsBlock(player).orElse(targetState.getBlock()).getPlacementState(quick);
+            if (res != null && correctObserverPlacement(targetState, res) && statesEqual(res, targetState)) {
+                contextCache = quick; quick.isAirPlace = true; return quick; }
         }
 
         for (Direction lookDirection : directionsToTry) {
-            final Direction side = Direction.UP; // Airplace only works for UP for now
-            BlockPos neighborPos = state.blockPos.offset(side);
-
-            Vec3d hitVec = Vec3d.ofCenter(state.blockPos);
-
-            BlockHitResult hitResult = new BlockHitResult(hitVec, side.getOpposite(), neighborPos, false);
-            PrinterPlacementContext context = new PrinterPlacementContext(player, hitResult, requiredItem, slot, lookDirection, false);
-            BlockState result = getRequiredItemAsBlock(player)
-                    .orElse(targetState.getBlock())
-                    .getPlacementState(context); // FIXME torch shift clicks another torch and getPlacementState is the clicked block, which is true
+            for (Direction side : candidateSides) {
```

Comment:
Not too sure about this implementation. I'll look some more at it and try to improve the code.